lab/lablib/midi/parse_midi.js

```diff
@@ -0,0 +1,279 @@
+    for (let i = 0; i < length; i++) {
+      data[i] = payload.getUint8(offset + 3 + i);
+    }
+
+    console.warn(
+      "detected unsupported MIDI meta event",
+      "0x" + meta_type.toString(16)
+    );
+    message = new MIDIOtherMeta(tick_delta, meta_type, data);
+  }
+
+  return [message, offset + META_HEADER_LENGTH + length];
+}
+
+function parse_sysex_message(payload, offset, tick_delta) {
+  const [length, next_offset] = parse_variable_length(payload, offset + 1);
+
+  const data = new Uint8Array(length);
+  for (let i = 0; i < length; i++) {
+    data[i] = payload.getUint8(next_offset + i);
+  }
+
+  return [new MIDISysex(tick_delta, data), next_offset + length];
+}
+
+/**
+ * Parse a MIDI message
+ * @param {DataView} payload the binary data of the track payload
+ * @param {number} offset the start offset for the body of this message (after the tick delta)
+ * @param {number} tick_delta tick delta for this message
+ * @returns {[import("./MidiFile.js").MIDIEvent, number]} (message, next_offset)
+ */
+function parse_midi_message(payload, offset, tick_delta) {
+  const status = payload.getUint8(offset);
+  const message_type = status >> 4;
+  const channel = status & 0xf;
+
+  if (
+    message_type === MIDIMessageType.NOTE_ON ||
+    message_type === MIDIMessageType.NOTE_OFF
+  ) {
+    const pitch = payload.getUint8(offset + 1);
+    const velocity = payload.getUint8(offset + 2);
+    const note = new MIDINote(
+      tick_delta,
+      channel,
+      message_type,
+      pitch,
+      velocity
+    );
+    return [note, offset + 3];
+  }
+
+  throw new Error(`not yet implemented: message type ${message_type}`);
+}
+
+/**
+ * Parse a MIDI track
+ * @param {MIDIChunk} chunk
+ * @returns {MIDITrack} The parsed track
+ */
+function parse_track(chunk) {
+  if (chunk.chunk_type !== "MTrk") {
+    throw new Error(`Invalid track chunk type ${chunk.chunk_type}`);
+  }
+
+  const events = [];
+  const payload = chunk.payload;
+  let offset = 0;
+  while (offset < payload.byteLength) {
+    const [tick_delta, next_offset] = parse_variable_length(payload, offset);
+    offset = next_offset;
+
+    // Take a look at the next byte to determine how to parse it
+    const peek = payload.getUint8(offset);
+    if (peek === META_MESSAGE) {
+      const [message, next_offset] = parse_meta_message(
+        payload,
+        offset,
+        tick_delta
+      );
+      offset = next_offset;
+      events.push(message);
+    } else if (peek === SYSEX_MESSAGE || peek === SYSEX_ESCAPE) {
+      parse_sysex_message(payload);
+    } else {
+      const [message, next_offset] = parse_midi_message(
+        payload,
+        offset,
+        tick_delta
+      );
+      offset = next_offset;
+      events.push(message);
+    }
+  }
+
+  return new MIDITrack(events);
+}
+
+/**
+ * Parse a MIDI file
+ * @param {ArrayBuffer} midi_buffer The MIDI file bytes as an ArrayBuffer
+ */
+export function parse_midi_file(midi_buffer) {
```

Comment:
[This article](https://midimusic.github.io/tech/midispec.html#BMA2_) gives a few simple MIDI files. Turn this into a set of unit tests.